cpp/oneapi/dal/backend/primitives/reduction/reduction.cpp

```diff
@@ -0,0 +1,145 @@
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include <type_traits>
+
+#include "oneapi/dal/backend/primitives/reduction/reduction.hpp"
+#include "oneapi/dal/backend/primitives/reduction/reduction_rm_rw.hpp"
+#include "oneapi/dal/backend/primitives/reduction/reduction_rm_cw.hpp"
+
+namespace oneapi::dal::backend::primitives {
+
+#ifdef ONEDAL_DATA_PARALLEL
+
+template <class Float, ndorder Layout, class BinaryOp, class UnaryOp>
+sycl::event reduce_rm_rw(sycl::queue& q,
+                         ndview<Float, 2, Layout>& input,
+                         ndview<Float, 1>& output,
+                         const BinaryOp binary,
+                         const UnaryOp unary,
+                         const event_vector& deps) {
+    using kernel_t = reduction_rm_rw<Float, BinaryOp, UnaryOp>;
+    const auto width = input.get_dimension(0);
```

Comment:
Yesterday it was discussed and it was decided that dimension(0) is for number of rows. I guess we need a strict rule for that.

Reply:
Fixed